### DISCUSSION TITLE
[GSoC 2025] Add framework for adding new drivers at runtime, for instance for cloud

### DISCUSSION THREAD

**Discussion by afbjorklund** [lima-vm/lima#2007] [Ideas]
The `docker-machine` program had a plugin system, so that you could provide new drivers besides the built-in ones.

This would be a good option to provide drivers for various proprietary hypervisors and cloud vendors, for instance.

They typically come with a big SDK, that might require certain hardware and large downloads when building.

So it would be nice if those were kept **out** of the Lima code base, but is still a nice addition to enable them...

----

`docker-machine` drivers:

https://github.com/machine-drivers/docker.github.io/blob/173d3c65f8e7df2a8c0323594419c18086fc3a30/machine/AVAILABLE_DRIVER_PLUGINS.md

https://github.com/machine-drivers/docker.github.io/blob/173d3c65f8e7df2a8c0323594419c18086fc3a30/machine/DRIVER_SPEC.md

Example of Lima drivers:

* https://github.com/lima-vm/lima/issues/1287 (PR #1277)

* https://github.com/lima-vm/lima/issues/2031 (PR #2054)

* vfkit driver (TBD)

* cloud drivers (TBD)

---

EDIT: Some general updates, after discussion below and feedback:

* The example drivers were just examples, and it might not be `vfkit` as a "new" driver and which cloud was not specified.
Note that the actual driver implementation is **not** part of the summer project, which is about the RPC plugin framework.
* The main code should not call directly into the driver code. This is supposed to be cleaned up, before GSoC project starts.
Note that this is more of a general rule for drivers (only SDK), but fixing other lima code is **not** in the project GSoC itself.
  * #1746
* Driver authors and implementations should not have to know about the RPC implementation, but only the Go interface.
* Drivers can be compiled in to the main lima applications or built as standalone executable, from the same driver code*.
* Changing which drivers are included in the main download, and which one is default for each os/arch, doesn't change.
* It should be possible to add new drivers _after_ release, as long as they use the same SDK version (i.e. the interface code)

Other details can be discussed on Slack or in other GitHub issues.

----

\* Here is another example, from a related project (containerd):
https://github.com/containerd/fuse-overlayfs-snapshotter/#option-1-embed-fuse-overlayfs-plugin-into-the-containerd-binary
https://github.com/containerd/fuse-overlayfs-snapshotter/#option-2-execute-fuse-overlayfs-plugin-as-a-separate-binary

> Two installation options are supported:
>
> 1. Embed fuse-overlayfs plugin into the containerd binary
> 2. Execute fuse-overlayfs plugin as a separate binary
>
> Choose 1 if you don't mind recompiling containerd, otherwise choose 2.

This is the same setup, as in docker-machine (mentioned above)

**Comment by jandubois** [comment]
Having pluggable drivers might be useful, but they should be maintained in separate repos (like `socket_vmnet`).

Random memory: I think `docker-machine-driver-hyperkit` must be run with as root and has several easily exploitable vulnerabilities (e.g. because it spawns subprocesses that also run as `root`). So this might be something to look out for as you implement other drivers.

**Reply by afbjorklund** [reply]
I was thinking to only add the API, but keep all (both?) supported drivers in-tree

**Reply by afbjorklund** [reply]
Also for me I think that qemu is plenty, it was mostly for the cloud...

**Reply by jandubois** [reply]
> but keep all (both?) supported drivers in-tree

Yes, the supported drivers need to stay in, off course, but things like adding a hyperkit driver, or cloud drivers, should be external. I don't actually see that anyone would actively maintain them, if the history of docker-machine drivers is any indication.

**Reply by afbjorklund** [reply]
What I meant was that I would not write another xhyve/hyperkit driver (especially not with vz already being there), I think we do need a hyperv driver but that can be done by `exec` to some PowerShell scripts rather than linking with a big SDK and thus it can stay in.

**Reply by afbjorklund** [reply]
I will do some kind of proof-of-concept interface next month, with one private cloud and one public cloud implementation.

**Comment by afbjorklund** [comment]
A simplified approach, for drivers that are internal to the program but have a lot of dependencies and bloat the binary, is to use the `plugin` interface instead of the `net/rpc`. The two major drawbacks is that it needs to be compiled at the same time as the program, and that it currently does not work on Windows (no .dll support yet, only .so). But it makes it easier to include something like libvirt, that has a lot of dependencies but allows for remote communication (the `qemu:///session` doesn't really add anything).

* https://pkg.go.dev/plugin

* https://pkg.go.dev/net/rpc

It would **not** be a solution for random cloud vendor drivers, that would be shipped independent of the `limactl` program. 

Those _would_ need some kind of RPC solution, perhaps wrapped in a framework like https://github.com/hashicorp/go-plugin

**Reply by AkihiroSuda** [reply]
Can we use REST or gRPC over stdio?

**Reply by afbjorklund** [reply]
I think that go-plugin can use both rpc and grpc.

**Reply by afbjorklund** [reply]
Having to define a public interface and a protocol for something like REST or gRPC is overkill for this use case, I think...

That was the benefit of using a "plugin" wrapper, over doing it ourselves. Just doing function or remote calls, like above.

**Reply by afbjorklund** [reply]
My attempt to load libvirt dynamically failed, due to it calling pkg-config (to find libvirt.pc) already from the go.mod...

The building and loading of the .so worked fine, but it was hard to make the plugin optional without a separate module.

**Comment by balajiv113** [comment]
Yes, during the initial driver interface as well the idea was to provide support via external plugin mechanism on future.

We should be easily able to expose a external plugin via https://pkg.go.dev/plugin with current code. Will just need driverutil to load the external drivers.

**Reply by afbjorklund** [reply]
If we use `plugin`, then the driver is not "external" anymore. It needs to be built at the same time, or it won't load later.

So if we want external plugins, then they will need to use RPC. Also, they were unable to module library dependencies.

**Reply by afbjorklund** [reply]
When doing the RPC driver, we should also consider how to forward configuration. This needs to be communicated as well.

We should **not** do it like minikube, when they either compile with the driver config or duplicate the struct without an API.

Note that we might not know (in the lima main program) what configuration that the plugin is added (in the plugin program).

For docker-machine, there was a rather hacky framework based on CLI flags which "worked" but json/yaml would be better...

**Reply by afbjorklund** [reply]
We also might have to add some new functions to the interface (=API), for instance the ability to upload images.

And how to send credentials to the driver, for instance passing on a token. Monitoring might also need revising.

**Comment by saz97** [comment]
Hello everyone,

I came across this discussion through the GSoC project description (https://github.com/cncf/mentoring/blob/main/programs/summerofcode/2025.md#vm-plugin-subsystem) and found the idea very interesting. I would love to join the discussion and contribute to this effort.

To get started, I plan to create a small demo that separates the QEMU driver into an independent binary that communicates with the main Lima binary via RPC (probably gRPC). 

I appreciate any guidance or suggestions from the maintainers and the community. Looking forward to contributing!

**Reply by afbjorklund** [reply]
That sounds like almost the entire project, and might be a bit big to start out with... Maybe just some simple RPC first, I would wait with gRPC (even if Lima has some helpers) until the actual project - but it doesn't hurt to learn more about it and protobuf

**Reply by saz97** [reply]
Thank you so much for your valuable advice!  I'll focus on creating a demo using simple RPC first and will explore gRPC and protobuf as I progress. thanks for your guidance again!

**Reply by saz97** [reply]
I have created an RPCDriver as the client-side of an RPC system. It starts a local RPC server by executing an external binary and establishes a connection. Could you please help review if there are any issues or suggest improvements?

Here are the links for reference:

PR: https://github.com/saz97/lima/pull/1/files
My branch: https://github.com/saz97/lima/tree/qemu-experiment
I’m a bit uncertain about my next steps: should I 

> switch the RPC to gRPC, 

> or continue with the existing RPC, and implement the QEMU driver within this RPC system?

**Reply by AkihiroSuda** [reply]
Thanks @saz97 


> PR

- `net/rpc` has been frozen and discouraged for new applications https://github.com/golang/go/issues/16844
- The socket should be listened on either UNIX socket or on stdio, not on TCP due to security concerns (even for 127.0.0.1). An example to use stdio can be found here: https://github.com/moby/buildkit/blob/v0.20.1/frontend/gateway/grpcclient/client.go#L1240
- Probably there should not be `--vm-type=rpc`. `--vm-type=${PLUGIN}` should look up a plugin in `${PREFIX}/libexec/lima/plugins/vm/${PLUGIN}` (or maybe just `lima-plugin-vm-${PLUGIN}` in the `${PATH}`) 
- All the QEMU-specific codes should be eventually decoupled from `pkg/start`, `pkg/limayaml`, and all other places: https://github.com/lima-vm/lima/issues/1746


> next steps

Please submit your application via https://summerofcode.withgoogle.com (Mar 25-Apr 9).

Applications will be reviewed and ranked through CNCF.
Timeline: https://developers.google.com/open-source/gsoc/timeline

**Reply by saz97** [reply]
Thanks for the feedback!

I understand that `net/rpc` is frozen and not recommended for new applications. This is just an initial version of demo, and I will update my code based on your suggestions, including switching to UNIX socket or stdio instead of TCP.

I got the point about _--vm-type=rpc_ not being a valid option. However, I’m a bit confused about the plugin lookup mechanism. Combining your guideline and  what I understand, there should be multiple plugin drivers in _lima/plugins/vm/_, such as _lima/plugins/vm/qemu, lima/plugins/vm/vz_, etc. The RPC server should act as a bridge, handling client requests and delegating them to the appropriate plugin driver. My mistake was adding a static `vmType=rpc`, whereas the valid vmType should be dynamically determined based on the available plugin drivers in _lima/plugins/vm/_. Please let me know if I got this right!

For these code should be decoupled from _pkg/start, pkg/limayaml_, and other modules. No problem, but i prefer this is a work beyond initial demo, maybe until the develop duration in project. but I try to work on this now because this sounds interesting!

Again, thanks for the guidance! Looking forward to your thoughts.

**Reply by AkihiroSuda** [reply]
> there should be multiple plugin drivers in lima/plugins/vm/, such as lima/plugins/vm/qemu, lima/plugins/vm/vz, etc. 

Right.

The vmType should be specified via `vmType` in the YAML or `--vm-type` in the CLI.
When the vmType is not specified, Lima should check `ok, err := getPlugin("vz").AcceptsTemplate(yaml)` and then falls back to the "qemu" plugin.

**Reply by saz97** [reply]
Hi everyone, I am working on integrating an RPC client into the QEMU driver to address the issue:

> Probably there should not be --vm-type=rpc. --vm-type=${PLUGIN} should look up a plugin in ${PREFIX}/libexec/lima/plugins/vm/${PLUGIN} (or maybe just lima-plugin-vm-${PLUGIN} in the ${PATH})

The updated PR link is https://github.com/saz97/lima/pull/1/files#diff-90164061f3572f8e2aca979d03cdd74187cbccfd16c2e9c28d349cbdf454155b. And the logs from both the RPC client and server are shown in the attached figure:
![image](https://github.com/user-attachments/assets/36d78122-23df-46df-9518-23ce9e2e0caa)

I am currently trying to understand the architecture of the RPC implementation. Specifically, I am unsure which parts of the original driver should be moved to the RPC server. In the original `qemu_driver `code, the QEMU process is started with the following command:
 `if err := qCmd.Start(); err != nil {
		return nil, err
	}`
If I move this logic to the RPC server, then the QEMU process will run there, while some setup tasks—such as downloading the nerdctl archive and pulling the image—will still be handled on the client side.

I would appreciate any guidance on whether my understanding is correct and what parts of the original implementation should be moved to the server. Thank you!

**Reply by AkihiroSuda** [reply]
> while some setup tasks—such as downloading the nerdctl archive and pulling the image—will still be handled on the client side.

Yes, correct, as these parts are not specific to the VM drivers.

**Reply by afbjorklund** [reply]
As mentioned, we might need a new API method to _upload_ images - but that can be added to the framework afterwards...

The current drivers just assume that they can copy files from the cache to the instance directory, without any extra steps.

**Comment by AkihiroSuda** [comment]
Cross-posting an announcement here for visibility:
- https://github.com/lima-vm/lima/discussions/3383

- - -

> This year we participate in the [Google Summer of Code](https://summerofcode.withgoogle.com)(GSoC) as a CNCF project, so as to facilitate contributions from open source beginners and students.
> 
> Successful contributors will receive [stipends](https://developers.google.com/open-source/gsoc/help/student-stipends).
> 
> **Register to become a contributor by April 8 (Tue) 18:00 UTC, via the [GSoC website](https://summerofcode.withgoogle.com)**.
> 
> (...)
